pkg/treeoperator/controller/virtualcluster_join_controller.go

```diff
@@ -1,16 +1,452 @@
+
+	kubeconfigStream, err := base64.StdEncoding.DecodeString(vc.Spec.Kubeconfig)
+	if err != nil {
+		return fmt.Errorf("decode target kubernetes kubeconfig %s err: %v", vc.Spec.Kubeconfig, err)
+	}
+	err, kosmosClient, _, _ := c.InitTargetKubeclinet(kubeconfigStream)
+	if err != nil {
+		return fmt.Errorf("create kubernetes client failed: %v", err)
+	}
+
+	old, err := kosmosClient.KosmosV1alpha1().Clusters().Get(context.TODO(),
+		clusterName, metav1.GetOptions{})
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("get cluster %s failed when we try to del: %v", clusterName, err)
+		}
+	} else {
+		err = c.RemoveClusterFinalizer(old, kosmosClient)
+		if err != nil {
+			return fmt.Errorf("removefinalizer %s failed: %v", clusterName, err)
+		}
+		err = kosmosClient.KosmosV1alpha1().Clusters().Delete(context.TODO(), clusterName, metav1.DeleteOptions{})
+		if err != nil {
+			return fmt.Errorf("delete cluster %s failed: %v", clusterName, err)
+		}
+	}
+	klog.Infof("Cluster %s has been deleted.", clusterName)
+	if err := c.RemoveFinalizer(ctx, vc); err != nil {
+		return fmt.Errorf("remove finalizer error: %v", err)
+	}
+	return nil
+}
+
+func (c *VirtualClusterJoinController) InitTargetKubeclinet(kubeconfigStream []byte) (error, versioned.Interface, kubernetes.Interface, extensionsclient.Interface) {
+	//targetKubeconfig := path.Join(DefaultKubeconfigPath, "kubeconfig")
+	//config, err := utils.RestConfig(targetKubeconfig, "")
+	config, err := utils.NewConfigFromBytes(kubeconfigStream)
+	if err != nil {
+		return fmt.Errorf("generate kubernetes config failed: %s", err), nil, nil, nil
+	}
+
+	kosmosClient, err := versioned.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("generate Kosmos client failed: %v", err), nil, nil, nil
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("generate K8s basic client failed: %v", err), nil, nil, nil
+	}
+
+	k8sExtensionsClient, err := extensionsclient.NewForConfig(config)
+	if err != nil {
+		return fmt.Errorf("generate K8s extensions client failed: %v", err), nil, nil, nil
+	}
+
+	return nil, kosmosClient, k8sClient, k8sExtensionsClient
+}
+
+func (c *VirtualClusterJoinController) DeployKosmos(ctx context.Context, request reconcile.Request, vc *v1alpha1.VirtualCluster) error {
+	klog.Infof("Start creating kosmos-clustertree secret")
+	clustertreeSecret, err := kosmosctl.GenerateSecret(treemanifest.ClusterTreeClusterManagerSecret, treemanifest.SecretReplace{
+		Namespace:  request.Namespace,
+		Cert:       cert.GetCrtEncode(),
+		Key:        cert.GetKeyEncode(),
+		Kubeconfig: vc.Spec.Kubeconfig,
+		Name:       request.Name,
+	})
+	if err != nil {
+		return err
+	}
+	err = c.Create(ctx, clustertreeSecret)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return fmt.Errorf("install clustertree error, secret created failed: %v", err)
+		}
+	}
+	klog.Info("Secret has been created. ")
```

Comment:
The log can be more detailed: please indicate which component the secret belongs to.

Reply:
add some log info